internal/integration/unified/client_entity.go

```diff
@@ -326,8 +354,47 @@ func (c *clientEntity) failedEvents() []*event.CommandFailedEvent {
 	return events
 }
 
-func (c *clientEntity) poolEvents() []*event.PoolEvent {
-	return c.pooled
+// filterEventsBySeq filters events by sequence number for the given eventType.
+// See comments on eventSequencer for more details.
+func filterEventsBySeq[T any](c *clientEntity, events []T, eventType monitoringEventType) []T {
+	cutoff := c.eventSequencer.cutoff.Load()
+	if cutoff == 0 {
+		return events
+	}
+
+	// Lock order: eventProcessMu -> eventSequencer.mu (matches writers)
+	c.eventProcessMu.RLock()
+	c.eventSequencer.mu.RLock()
+
+	// Snapshot to minimize time under locks and avoid races
+	localEvents := append([]T(nil), events...)
+
+	var seqSlice []int64
+	if eventType == poolAnyEvent {
+		seqSlice = c.eventSequencer.poolSeq
+	} else {
+		seqSlice = c.eventSequencer.seqByEventType[eventType]
+	}
+
+	localSeqs := append([]int64(nil), seqSlice...)
```

Comment:
Optional: Use `slices.Clone`.

```suggestion
	localSeqs := slices.Clone(seqSlice)
```